### PR TITLE
ABCL-STEPPER:  implementation of CL:STEP for interpreted code

### PR DESCRIPTION
**Pull request by alejandrozf**
# Working stepper for ABCL now can run in Slime!

This PR was made on top of PR [568](https://github.com/armedbear/abcl/pull/568) and brings to the table the possibility to run the stepper also inside the main REPL in Slime :100: ! (tested on Slime commit 50d4a7b1). 

It is also rebased with the latest changes in ABCL from master branch.

# See also 

Other possible strategies for implementation <<https://github.com/armedbear/abcl/issues/264>>.

### CODE REVIEW

**Comment by easye** [comment]
@alejandrozf Congratulations on getting SLIME to work!  I have been spending my recent efforts on deprecating the use of ABCL specialized Slime{Input,Output}Stream <<https://github.com/armedbear/abcl/pull/586>>.  I intend to release abcl-1.9.2 <<https://github.com/armedbear/abcl/issues/594>> "real soon now", for which I will review your stepper for inclusion, so expect feedback early next week.  

best,
Mark

**Comment by alejandrozf** [comment]
@easye 
Thank you, Mark! :) 

Yes, I just tested it for the Slime version v2.28 (commit a4f34714) and it is working fine too

File: src/org/armedbear/lisp/Lisp.java [hunk 1]
```diff
@@ -54,6 +54,72 @@
 
   public static boolean initialized;
 
+  // stepping related attributes
```
Comment (easye) [inline src/org/armedbear/lisp/Lisp.java:57]: @alejandrozf Please move all the "new" code for the stepper to the end of `Lisp.java`.  Code that changes existing functions is fine to modify "in-place".  

This non-functional change will help in looking through the history of `Lisp.java` with tools that compare lines directly without additional heuristics (like `patch` and `diff`).

File: src/org/armedbear/lisp/SocketStream.java [hunk 2]
```diff
@@ -47,6 +47,10 @@ public SocketStream(Socket socket, Stream in, Stream out)
         this.socket = socket;
     }
 
+    public Socket getJavaSocket() {
```
Comment (easye) [inline src/org/armedbear/lisp/SocketStream.java:50]: I don't have a problem making this interface public, but please separate out into a different pull request.

File: src/org/armedbear/lisp/Lisp.java [hunk 3]
```diff
@@ -2286,6 +2492,136 @@ public LispObject execute(LispObject key, LispObject value)
       }
     };
 
+  // ### %set-stepping-task-on
```
Comment (easye) [inline src/org/armedbear/lisp/Lisp.java:2495]: A more "Lisp-y" way might be to implement `defsetf-expanders` on various specials, but since these reader/writers are essentially direct toggles of Java booleans, I guess this is fine.

File: src/org/armedbear/lisp/Lisp.java [hunk 4]
```diff
@@ -517,6 +583,73 @@ public static final LispObject eval(LispObject obj)
...
+      funcall(functionPrintStepperStr, new LispObject[] {
+          new SimpleString("step " + stepNumberInternal + " ==> value: " + result.printObject()),
+          Symbol.T
+        },
+        thread);
+    }
+    thread._values = values;
+  }
+
   public static final LispObject eval(final LispObject obj,
```
Comment (easye) [inline src/org/armedbear/lisp/Lisp.java:653]: Have you done any performance benchmarks?  We are adding additional steps to the Lisp.eval() codepath here, even if we aren't stepping through functions.  From looking at this code, it seems like the only "penalty" is in the boolean compares for `stepInSymbolResult` et. al. but it would be nice to confirm there isn't a generic speed penalty here.

**Review by easye — CHANGES_REQUESTED** [review]
This is more or less ready to commit "as is" modulo reordering the code in `Lisp.java` as far as I am concerned.  

Providing a reference to your 2023 paper would be good as well.

**Comment by easye** [comment]
@alejandrozf Feel free to edit the commit history under this branch (i.e. "git force push") to get the code in a good state.  I can help you squash/edit/reorder commits for you as well.

**Comment by alejandrozf** [comment]
Thank you @easye !

I'll be working on it
